config/coordinator.go

```diff
@@ -0,0 +1,159 @@
+
+	// Protects config and subscribers
+	mutex       sync.Mutex
+	config      *Config
+	subscribers []func(*Config) error
+
+	configHashMetric        prometheus.Gauge
+	configSuccessMetric     prometheus.Gauge
+	configSuccessTimeMetric prometheus.Gauge
+}
+
+// NewCoordinator returns a new coordinator with the given configuration file
+// path. It does not yet load the configuration from file. This is done in
+// `Reload()`.
+func NewCoordinator(configFilePath string, r prometheus.Registerer, l log.Logger) *Coordinator {
+	c := &Coordinator{
+		configFilePath: configFilePath,
+		logger:         l,
+	}
+
+	c.registerMetrics(r)
+
+	return c
+}
+
+func (c *Coordinator) registerMetrics(r prometheus.Registerer) {
+	configHash := prometheus.NewGauge(prometheus.GaugeOpts{
+		Name: "alertmanager_config_hash",
+		Help: "Hash of the currently loaded alertmanager configuration.",
+	})
+	configSuccess := prometheus.NewGauge(prometheus.GaugeOpts{
+		Name: "alertmanager_config_last_reload_successful",
+		Help: "Whether the last configuration reload attempt was successful.",
+	})
+	configSuccessTime := prometheus.NewGauge(prometheus.GaugeOpts{
+		Name: "alertmanager_config_last_reload_success_timestamp_seconds",
+		Help: "Timestamp of the last successful configuration reload.",
+	})
+
+	r.MustRegister(configHash, configSuccess, configSuccessTime)
+
+	c.configHashMetric = configHash
+	c.configSuccessMetric = configSuccess
+	c.configSuccessTimeMetric = configSuccessTime
+}
+
+// Subscribe subscribes the given Subscribers to configuration changes.
+func (c *Coordinator) Subscribe(ss ...func(*Config) error) {
+	c.mutex.Lock()
+	defer c.mutex.Unlock()
+
+	c.subscribers = append(c.subscribers, ss...)
+}
+
+func (c *Coordinator) notifySubscribers() error {
+	for _, s := range c.subscribers {
+		if err := s(c.config); err != nil {
+			return err
+		}
+	}
+
+	return nil
+}
+
+// loadFromFile triggers a configuration load, discarding the old configuration.
+func (c *Coordinator) loadFromFile() error {
+	level.Info(c.logger).Log(
+		"msg", "Loading configuration file",
+		"file", c.configFilePath,
+	)
+
+	conf, plainConfig, err := LoadFile(c.configFilePath)
+	if err != nil {
+		c.configSuccessMetric.Set(0)
+		level.Error(c.logger).Log(
+			"msg", "Loading configuration file failed",
+			"file", c.configFilePath,
+			"err", err,
+		)
+		return err
+	}
+
+	c.config = conf
+	c.configSuccessMetric.Set(1)
+	c.configSuccessTimeMetric.Set(float64(time.Now().Unix()))
+	hash := md5HashAsMetricValue(plainConfig)
+	c.configHashMetric.Set(hash)
+
+	return nil
+}
+
+// Reload triggers a configuration reload from file and notifies all
+// configuration change subscribers.
+func (c *Coordinator) Reload() error {
+	c.mutex.Lock()
+	defer c.mutex.Unlock()
+
+	err := c.loadFromFile()
+	if err != nil {
+		c.logger.Log(
+			"msg", "loading configuration file failed",
+			"file", c.configFilePath,
+			"err", err,
+		)
+		return err
+	}
+
+	err = c.notifySubscribers()
+	if err != nil {
```

Comment:
any reason for not having the `if err := foo.Thing(); err != nil` pattern?

Reply:
Oh, right. That looks better. Adjusted above as well.